File: slot-dart/lib/src/views/reel_view.dart
library slot.views.reel_view;

import 'package:stagexl/stagexl.dart';

import 'package:slot/src/views/icon_view.dart';
import 'package:slot/src/models/reel_model.dart';
import 'package:slot/src/mvc/view.dart';
import 'package:slot/src/core/engine.dart';

class ReelView extends View {

  List<IconView> _iconViews;
  num _tweenCount;
  num _stopPosition;
  Function _onStopped;

  ReelView(Model model) : super(model) {

    _iconViews = new List<IconView>();
    _tweenCount = 0;

    _setupInitialIcons();

  }

  void stop(num position) {
    _stopPosition = model.indexToReelPosition(position);
  }

  void spin() {
    shiftReel();
  }

  void shiftReel() {

    // Removing bottom out of view icon.
    removeChild(_iconViews.removeLast());

    // Moving the position of the reel one forward.
    num position = model.position;
    position++;

    // Reseting the reel position back to the start if new position is out of bounds.
    if (position == model.iconModels.length) {
      position = 0;
    }

    model.position = position;

    IconModel newIconModel = model.getIconModel(-2 + model.position);
    IconView newIconView = new IconView(newIconModel);

    _iconViews.insert(0, newIconView);

    _tweenCount = _iconViews.length - 1;
    for (num i = 1; i < _iconViews.length; i++) {
      IconView iconView = _iconViews[i];
      Tween tween = new Tween(iconView, 0.1, TransitionFunction.linear);
      tween.onComplete = _onShiftedReel;
      tween.animate.y.to(iconView.y + iconView.height);
      Engine.game.stage.juggler.add(tween);
    }

    addChild(newIconView);

  }

  void _onShiftedReel() {
    _tweenCount--;
    if (_tweenCount == 0) {
      // Stop spinning condition.
      if (model.position == _stopPosition) {
        stopReel();
      } else {
        shiftReel();
      }
    }
  }

  void stopReel() {
    _stopPosition = null;
    if (_onStopped) {
      {
        _onStopped.apply(this);
      }
    }
  }

  void _setupInitialIcons() {
    for (var i = -2; i <= 2; ++i) {
      IconView iconView = new IconView(model.getIconModel(i + model.position));
      addChild(iconView);
      iconView.y = (i + 2) * iconView.height;
      _iconViews.add(iconView);
    }
  }

  ReelModel get model {
    return super.model;
  }

  void set onStopped(void function()) {
    _onStopped = function;
  }

}

File: slot-ts/src/views/ReelView.ts
import Actuate from 'motion/Actuate';
import Linear from 'motion/easing/Linear';
import View from '../engine/mvc/View';
import IconModel from '../models/IconModel';
import ReelModel from '../models/ReelModel';
import IconView from './IconView';

export default class ReelView extends View {
  private iconsViews: IconView[] = [];
  private tweenCount: number = 0;
  private stopPosition: number = NaN;
  public onStopped: Function | null = null;

  constructor(model: ReelModel) {
    super(model);

    this.iconsViews = [];
    this.tweenCount = 0;

    this.setupInitialIcons();
  }

  public stop = (stopPosition: number): void => {
    this.stopPosition = this.getModel().getLoopingIndex(stopPosition);
  }

  public spin = (): void => {
    this.shiftReel();
  }

  private shiftReel = (): void => {
    if (this.iconsViews) {
      // Removing last icon at the bottom of the reel.
      const iconView = this.iconsViews.pop();
      if (iconView) {
        this.removeChild(iconView);
      }

      // Moving the position of the reel one forward.
      let position: number = this.getModel().getPosition();
      position++;

      // Reseting the reel position back to the start if new position is out of bounds.
      if (position === this.getModel().getIconModels().length) {
        position = 0;
      }

      // Apply the new limited position to the reel.
      this.getModel().setPosition(position);

      // Adding the next icon at the top of the reel.
      const newIconModel: IconModel = this.getModel().getIconModel(
        -2 + this.getModel().getPosition()
      );
      const newIconView: IconView = new IconView(newIconModel);
      this.iconsViews.unshift(newIconView);

      this.tweenCount = this.iconsViews.length - 1;
      for (let i: number = 1; i < this.iconsViews.length; i++) {
        const iconView: IconView = this.iconsViews[i];
        Actuate.tween(iconView, 0.2, { y: iconView.y + iconView.height })
          .ease(Linear.easeNone)
          .onComplete(this.onShiftedReel);
      }
      this.addChild(newIconView);
    }
  }

  private onShiftedReel = (): void => {
    this.tweenCount--;
    if (this.tweenCount === 0) {
      // Stop spinning condition.
      if (this.getModel().getPosition() === this.stopPosition) {
        this.stopReel();
      } else {
        this.shiftReel();
      }
    }
  }

  private stopReel = (): void => {
    this.stopPosition = NaN;
    if (this.onStopped) {
      this.onStopped.apply(this);
    }
  }

  private setupInitialIcons = (): void => {
    for (let i: number = -2; i <= 2; i++) {
      const iconModel: IconModel = this.getModel().getIconModel(i + this.getModel().getPosition());
      const iconView: IconView = new IconView(iconModel);
      this.addChild(iconView);
      iconView.y = (i + 2) * iconView.height;
      this.iconsViews.push(iconView);
    }
  }

  public getIconView = (index: number): IconModel | null => {
    if (index < this.iconsViews.length) {
      return this.iconsViews[index].getModel();
    }
    return null;
  }

  protected onUpdated = (): void => {
    super.onUpdated();
  }

  getModel(): ReelModel {
    return this.model as ReelModel;
  }
}
